src/Uri.php

```diff
@@ -261,9 +261,9 @@ public function withUserInfo($user, $password = null)
             ));
         }
 
-        $info = $user;
+        $info = $this->filterUserInfoPart($user);
         if ($password) {
-            $info .= ':' . $password;
+            $info .= ':' . $this->filterUserInfoPart($password);
```

Comment:
On second thought, maybe we can change this to a BC compatible solution by _not_ using `self::CHAR_UNRESERVED` and `self::CHAR_SUB_DELIMS` for user info. Instead, we can/should/must assume that input is already urlencoded if it comes from the uri string, and can/should/must assume it's not if it is set via `withUserInfo` method.

```
        $info = urlencode($user);
        if ($password) {
            $info .= ':' . urlencode($password);
        }
```

Now, if that would be considered bc breaking, we could add a third optional arg to `withUserInfo`. E.g.

```
public function withUserInfo($user, $password = null, $urlEncoded = true)
    {
   // snip
```

But I'd argue that if someone is already urlencoding this input, they're working around the bug, and we shouldn't worry about BC.